src/.prototype/musicutils/ts/keysignature.ts

```diff
@@ -0,0 +1,1640 @@
+        'major blues': [2, 1, 1, 3, 2, 2],
+        'whole tone': [2, 2, 2, 2, 2, 2],
+        // 5 notes in an octave.
+        'major pentatonic': [2, 2, 3, 2, 3],
+        'minor pentatonic': [3, 2, 2, 3, 2],
+        'chinese': [4, 2, 1, 4, 1],
+        'egyptian': [2, 3, 2, 3, 2],
+        'hirajoshi': [1, 4, 1, 4, 2],
+        'in': [1, 4, 2, 1, 4],
+        'minyo': [3, 2, 2, 3, 2],
+        'fibonacci': [1, 1, 2, 3, 5]
+    };
+
+    /** These maqam mode names imply a specific key. */
+    static readonly MAQAM_KEY_OVERRIDES: { [key: string]: string } = {
+        'hijaz kar': 'c',
+        'hijaz kar maqam': 'c',
+        'shahnaz': 'd',
+        'maqam mustar': 'eb',
+        'maqam jiharkah': 'f',
+        'shadd araban': 'g',
+        'suzidil': 'a',
+        'ajam': 'bb',
+        'ajam maqam': 'bb'
+    };
+
+    static readonly MODE_MAPPER: { [key: string]: string } = {
+        'ionian': 'major',
+        'aeolian': 'minor',
+        'natural minor': 'minor',
+        'harmonic minor': 'minor',
+        'major': 'major',
+        'minor': 'minor'
+    };
+
+    /** These key signaturs (and their equivalents) prefer sharps over flats. */
+    static readonly PREFER_SHARPS: string[] = [
+        'c major',
+        'c major pentatonic',
+        'c major blues',
+        'c whole tone',
+        'd dorian',
+        'e phrygian',
+        'f lydian',
+        'g mixolydian',
+        'a minor',
+        'a minor pentatonic',
+        'b locrian',
+        'g major',
+        'g major pentatonic',
+        'g major blues',
+        'g whole tone',
+        'a dorian',
+        'b phrygian',
+        'c lydian',
+        'd mixolydian',
+        'e minor',
+        'e minor pentatonic',
+        'f# locrian',
+        'd major',
+        'd major pentatonic',
+        'd major blues',
+        'd whole tone',
+        'e dorian',
+        'f# phrygian',
+        'g lydian',
+        'a mixolydian',
+        'b minor',
+        'b minor pentatonic',
+        'c# locrian',
+        'a major',
+        'a major pentatonic',
+        'a major blues',
+        'a whole tone',
+        'b dorian',
+        'c# phrygian',
+        'd lydian',
+        'e mixolydian',
+        'f# minor',
+        'f# minor pentatonic',
+        'e major',
+        'e major pentatonic',
+        'e major blues',
+        'e whole tone',
+        'f# dorian',
+        'a lydian',
+        'b mixolydian',
+        'c# minor',
+        'c# minor pentatonic',
+        'b major',
+        'b major pentatonic',
+        'b major blues',
+        'b whole tone',
+        'c# dorian',
+        'd# phrygian',
+        'e lydian',
+        'f# mixolydian',
+        'g# minor',
+        'g# minor pentatonic',
+        'a# locrian'
+    ];
+
+    key: string;
+    mode: string;
+    scale: string[];
+    halfSteps: number[];
+    keySignature: string;
+    genericScale: string[];
+    scalarModeNumbers: string[];
+    eastIndianSolfegeNotes: string[];

```

Comment:
Which of these should we expect to be directly accessed through the instance, and which are private?

Also need docs.